### PR TITLE
Add size limit for loadX methods

### LINKED ISSUE #2981: show error message when loading files that are too big
#### Nature of issue?

- [ ] Found a bug
- [x] Existing feature enhancement
- [ ] New feature request

#### Most appropriate sub-area of p5.js?

- [ ] Color
- [ ] Core/Environment/Rendering
- [ ] Data
- [ ] Events
- [ ] Image
- [x] IO
- [ ] Math
- [ ] Typography
- [ ] Utilities
- [ ] WebGL
- [ ] Other (specify if possible)

#### Feature enhancement details:

There are currently two open issues for when the browser fails to load a file because it is too big: #2661 for `loadStrings()` and #2216 for `loadTable()`. I think they are not possible to fix, but it would be good to produce an error message for the user and suggest breaking the data into smaller files as a work around.

### PR DESCRIPTION
**Pull request by TanviKumar**
This is to close #2981 . Uses HEAD request to get the content-length of the file. If found to be greater than 64 MB, error is produced, else the file is fetched as before.

### CODE REVIEW

**Comment by almchung** [comment]
@TanviKumar Aiming for more of a warning than an error message would give a more suitable tone for this case, what do you think?  So maybe use phrases such as 'warning' or 'is recommended...'?

**Comment by TanviKumar** [comment]
@almchung I'll go ahead and update the terms used in the warning. Currently my code doesn't fetch the file if the size exceeds the limit. If we are to give the user just a warning, should I alter the code such that the file is fetched anyway but a warning is given if size is larger than expected?

File: src/io/files.js [hunk 1]
```diff
@@ -1085,6 +1088,68 @@ p5.prototype.httpDo = function() {
       }
     }
 
+    if (method === 'GET') {
+      // Get the size of the file to be fetched using HEAD request
+      return new Promise(function(resolve, reject) {
+        var xhr = new XMLHttpRequest();
```
Comment (lmccart) [inline src/io/files.js]: it looks like there are now two HTTP requests being made per httpDo call.. this one here and the request() on line 1153. is this needed? or is there a way to combine?

Comment (lmccart) [reply src/io/files.js]: and another one in getFile() called from line 1116?
I'm wondering if this function just needs to be cleaned up and streamlined so it's clear what's going on.
some inline comments would help a lot too.

File: src/io/files.js [hunk 2]
```diff
@@ -1093,8 +1158,72 @@ p5.prototype.httpDo = function() {
...
+        type = 'text';
+      }
+    }
+
+    if (type === 'jsonp') {
+      promise = fetchJsonp(path, jsonpOptions);
+    } else {
+      promise = fetch(request);
+    }
+    promise = promise.then(function(res) {
```
Comment (lmccart) [inline src/io/files.js]: this line feels weird to me, can you explain what's going on here?

File: src/io/files.js [hunk 3]
```diff
@@ -941,6 +941,7 @@ p5.prototype.httpPost = function() {
  * For more advanced use, you may also pass in the path as the first argument
  * and a object as the second argument, the signature follows the one specified
  * in the Fetch API specification.
+ * This method will only fetch files upto size of 64MB when "GET" is used.
```
Comment (lmccart) [inline src/io/files.js]: add space between "up" and "to"

File: src/io/files.js (hunk 1 above)
Comment (TanviKumar) [reply src/io/files.js]: If a file has to be loaded, and a GET request has to be made, first a HEAD request is made. The HEAD request only fetches the headers and doesn't fetch the entire file so it doesn't take much longer at all. It gives the file size which we use to produce a warning. A normal GET request is then made after the HEAD request is complete to fetch the file the user wants.

So for GET requests, 2 HTTP requests are made, the one in line 1094 and 1116. For other types of requests, just the one in 1116 is made.

File: src/io/files.js (hunk 3 above)
Comment (TanviKumar) [reply src/io/files.js]: Will do!

File: src/io/files.js (hunk 2 above)
Comment (TanviKumar) [reply src/io/files.js]: This is the same code as present before, I'll add another parameter to the getFile() method so the code isn't repeated twice.

**Comment by TanviKumar** [comment]
@lmccart I've pushed some code that makes things a lot cleaner and more modular

**Comment by meiamsome** [comment]
I think `fetch` should be used in preference to XHR - You could just duplicate the current `Request` but set the method to `HEAD`. In either case, that makes the minimum amount of requests for a GET to the same host 2 round-trips and if CORS is used in certain situations this may even be 4 (CORS will send an extra OPTIONS request ahead of time to check that the CORS request is valid before sending it, depending upon if the request is considered 'safe')

As a guess, it may be possible to just read the `Content-Length` header from the fetch response using something like `res.headers.get('Content-Length')` before the body is read, somewhere around here: https://github.com/processing/p5.js/blob/989a25b8b87d06914fd5d6c2a59baee6ca1a7baf/src/io/files.js#L1122
That's entirely untested speculation, though, and by the looks of a quick search you'd have to handle that that header may be `null` if it's unsent or blocked by CORS

**Comment by TanviKumar** [comment]
@meiamsome Will test out getting content length from the fetch response and get back to you.

For the case where no content length information is sent or available, I'm going to add a line about the known size limitations to the documentation of the loadX methods so users are informed beforehand

**Comment by TanviKumar** [comment]
@meiamsome Considering we'll now be fetching the file no matter what the result and only producing a warning, I've changed the code significantly based on your suggestions and also updated much of the documentation. @lmccart @almchung This should seem a lot neater!

**Comment by meiamsome** [comment]
This looks like a good solution to me. I think warning and continuing anyway is the best situation so as to avoid preventing people who are manually changing their browser's limits being restricted.

**Comment by TanviKumar** [comment]
@almchung Do let me know if any changes are needed!

**Comment by lmccart** [comment]
looks great, thanks @TanviKumar and @almchung and @meiamsome for reviewing!

**Pull request merged**